06/Coordinatron.scala

```diff
@@ -0,0 +1,109 @@
+import scala.io.Source
+
+object Coordinatron {
+  val bufferedSource = Source.fromFile("dataset")
+  val lines = bufferedSource.getLines.toList
+  bufferedSource.close
+
+  def absoluteValue(n: Int): Int = {
+    if (n < 0) n * -1 else n
+  }
+
+  def distance(p1: (Int, Int), p2: (Int, Int)): Int = {
+    absoluteValue(p1._1 - p2._1) + absoluteValue(p1._2 - p2._2)
+  }
+
+  def max(a: Int, b: Int): Int = {
+    if (a > b) a else b
+  }
+
+  def min(a: Int, b: Int): Int = {
+    if (a < b) a else b
+  }
+
+  // finds the bounds as follows: x min, x max, y min, y max
+  def bounds(l: List[(Int, Int)]): (Int, Int, Int, Int) = {
+    @annotation.tailrec
+    def go(l: List[(Int, Int)], xMin: Int, xMax: Int, yMin: Int, yMax: Int): (Int, Int, Int, Int) = {
+      if (l.isEmpty) (xMin, xMax, yMin, yMax)
+      else go(l.tail, min(l.head._1, xMin), max(l.head._1, xMax), min(l.head._2, yMin), max(l.head._2, yMax))
+    }
+    go(l.tail, l.head._1, l.head._1, l.head._2, l.head._2)
+  }
+
+  // returns the list's closest point (or one of them), and true if equidistant to multiple points or false otherwise
+  def findClosestPoint(p: (Int, Int), l: List[(Int, Int)]): ((Int, Int), Boolean) = {
+    def go(p: (Int, Int), l: List[(Int, Int)], s: Int, c: (Int, Int), e: Boolean): ((Int, Int), Boolean) = {
+      if (l.isEmpty) (c, e)
+      else {
+        val currentDistance: Int = distance(p, l.head)
+        val shortest: Int = min(currentDistance, s)
+        val closer: (Int, Int) = if (shortest == s) c else l.head
+        /* if shortest doesn't equal s, it changed; equi must become false
+         if currentDistance equals s, we have seen this distance before; equi must become true
+         in all other cases equi must remain at the value of e */
+        val equidistant: Boolean = if (shortest != s) false else if (currentDistance == s) true else e
+        go(p, l.tail, shortest, closer, equidistant)
+      }
+    }
+
+    go(p, l.tail, distance(p, l.head), l.head, false)
+  }
+
+
+
+  // do a loop from all bounds minus one, calculate each point distance, then be done
+  // also, save a list of infinites (closest points to outside border are hitting infinite space)
+  def makeMap(l: List[(Int, Int)], bounds: (Int, Int, Int, Int)): (Map[(Int, Int), Int], List[(Int, Int)]) = {
+    def updateMap(m: Map[(Int, Int), Int], p: ((Int, Int), Boolean)): Map[(Int, Int), Int] = {
+      if (p._2) m
+      else m.updated(p._1, m.get(p._1) match {
+          case Some(i) => i + 1
+          case None => 1
+        })
+    }
+
+    def updateList(l: List[(Int, Int)], p: ((Int, Int), Boolean)): List[(Int, Int)] = {
+      if (p._2 || l.contains(p)) l else p._1 :: l
```

Comment:
`l.contains(p)` will always return `false`, since `l` is a `List[(Int, Int)]` and `p` is a `((Int, Int), Boolean)`.